rasterio/io.py

```diff
@@ -195,6 +202,66 @@ def __repr__(self):
             self.closed and 'closed' or 'open', self.name, self.mode)
 
 
+class MemoryFile(MemoryFileBase):
+    """A BytesIO-like object, backed by an in-memory file.
+
+    This allows formatted files to be read and written without I/O.
+
+    A MemoryFile created with initial bytes becomes immutable. A
+    MemoryFile created without initial bytes may be written to using
+    either file-like or dataset interfaces.
```

Comment:
I'm not seeing where immutability is enforced in the Cython code.  Did I miss something obvious or is this handled by GDAL?